x/incentive/keeper/refundable_msg_index.go

```diff
@@ -1,31 +1,99 @@
 package keeper
 
 import (
+	"bytes"
 	"context"
+	"fmt"
+	"time"
 
-	"github.com/babylonlabs-io/babylon/v4/x/incentive/types"
+	errorsmod "cosmossdk.io/errors"
+	"cosmossdk.io/x/feegrant"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
-)
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/gogoproto/proto"
 
-// getRecipient returns the address that should receive the refund.
-func getRecipient(feeTx sdk.FeeTx) sdk.AccAddress {
-	if feeGranter := feeTx.FeeGranter(); feeGranter != nil {
-		return feeGranter
-	}
-	return feeTx.FeePayer()
-}
+	"github.com/babylonlabs-io/babylon/v4/x/incentive/types"
+)
 
-// RefundTx refunds the given tx by sending the fee back to the fee payer.
+// RefundTx refunds the transaction fee to the appropriate party.
+// If a fee grant was used, it restores the granted allowance.
+// Otherwise, it sends the fee back to the original fee payer.
 func (k Keeper) RefundTx(ctx context.Context, tx sdk.FeeTx) error {
 	txFee := tx.GetFee()
 	if txFee.IsZero() {
-		// not possible with the global min gas price mechanism
-		// but having this check for compatibility in the future
 		return nil
 	}
-	txFeePayer := getRecipient(tx)
 
-	return k.bankKeeper.SendCoinsFromModuleToAccount(ctx, k.feeCollectorName, txFeePayer, txFee)
+	feeGranter := tx.FeeGranter()
+	feePayer := tx.FeePayer()
+
+	if feeGranter != nil && !bytes.Equal(feeGranter, feePayer) {
+		return k.refundToFeeGranter(ctx, feeGranter, feePayer, txFee)
+	}
+
+	// refund to fee payer
+	return k.bankKeeper.SendCoinsFromModuleToAccount(ctx, k.feeCollectorName, feePayer, txFee)
+}
+
+// refundToFeeGranter restores the fee grant allowance by increasing the spend limit or restoring the deleted grant
+// and sends the refund to the fee granter
+func (k Keeper) refundToFeeGranter(ctx context.Context, feeGranter, feePayer sdk.AccAddress, refund sdk.Coins) error {
```

Comment:
this part got complex, could you create tests for this specific function with a real `feegrantKeeper` structure, not mock

Reply:
with expected allowance+balance checks after the `refundToFeeGranter` is called

Reply:
yes, will work on additional tests